openem-ingestor/core/taskqueue.go

```diff
@@ -0,0 +1,163 @@
+package core
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/wailsapp/wails/v2/pkg/runtime"
+)
+
+type TaskQueue struct {
+	datasetSourceFolders sync.Map           // Datastructure to store all the upload requests
+	inputChannel         chan IngestionTask // Requests to upload data are put into this channel
+	resultChannel        chan TaskResult    // The result of the upload is put into this channel
+	AppContext           context.Context
+	Config               Config
+}
+
+type TransferMethods int
+
+const (
+	TransferS3 TransferMethods = iota + 1
+	TransferGlobus
+)
+
+type TransferOptions struct {
+	S3_endpoint string
+	S3_Bucket   string
+	Md5checksum bool
+}
+
+type IngestionTask struct {
+	// DatasetFolderId   uuid.UUID
+	DatasetFolder     DatasetFolder
+	ScicatUrl         string
+	ScicatAccessToken string
+	TransferMethod    TransferMethods
```

Comment:
maybe somewhere here a function pointer could be put in that starts/does the file transfer, whose function signature is the same for all transfer types through the use of the TransferOptions struct. The underlying function would then only use those parameters in TransferOptions which are relevant to the transfer at hand.

Reply:
This part of actually implementing the ingestion/transfer I would leave to you. I can give you the options that are needed for S3 but you seem to have a good idea on how to abstract that.